File: plugins/kraken_bundle/ios/Classes/KrakenBundleManager.m
//
//  BundleManager.m
//  kraken_bundle
//
//  Created by lzl on 2020/3/30.
//

#import "KrakenBundleManager.h"

@implementation KrakenBundleManager;

static KrakenBundleManager * _instance;

+(instancetype)shareBundleManager{
    return [[self alloc]init];
}

+(instancetype)allocWithZone:(struct _NSZone *)zone{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _instance = [super allocWithZone:zone];
    });
    return _instance;
}

-(void)setUp:(NSString*)bundleUrl zipUrl:(NSString*)zipBundleUrl {
  if (bundleUrl != nil) {
    _instance->_bundleUrl = bundleUrl;
  }
  if (zipBundleUrl != nil) {
    _instance->_zipBundleUrl = zipBundleUrl;
  }
}

-(void)reload {
  if (_instance->_channel != nil) {
    [_instance->_channel invokeMethod:@"reload"
                            arguments:nil];
  }
}

-(void)onAttach:(KrakenBundlePlugin*)krakenBundlePlugin channel:(FlutterMethodChannel*) channel {
  _instance->_krakenBundlePlugin = krakenBundlePlugin;
  _instance->_channel = channel;
}

-(void)onDetach {
  _instance->_krakenBundlePlugin = nil;
}

-(NSString*)getBundleUrl {
  return _instance->_bundleUrl;
}

-(NSString*)getZipBundleUrl{
  return _instance->_zipBundleUrl;
}

- (nonnull id)copyWithZone:(nullable NSZone *)zone {
  return _instance;
}

- (nonnull id)mutableCopyWithZone:(nullable NSZone *)zone {
  return _instance;
}

@end

File: plugins/kraken_bundle/macos/Classes/KrakenBundleManager.swift
/*
* Copyright (C) 2019-present Alibaba Inc. All rights reserved.
* Author: Kraken Team.
*/

import Cocoa
import FlutterMacOS

public class KrakenBundleManager: NSObject {
  private static let instance:KrakenBundleManager = KrakenBundleManager()
  private var bundleUrl:String?
  private var zipBundleUrl:String?
  private var bundlePath:String?
  private var krakenBundlePlugin:KrakenBundlePlugin?
  private var channel:FlutterMethodChannel?
  
  private override init() {
    bundleUrl = nil
    zipBundleUrl = nil
    bundlePath = nil
  }
  
  public static var shared: KrakenBundleManager {
      return self.instance
  }
  /**
   * page load form
   * priority bundleUrl > zipBundleUrl
   * @param bundleUrl
   * @param zipBundleUrl
   */
  public func setUp(bundleUrl:String? = nil, zipBundleUrl:String? = nil, bundlePath:String? = nil) {
    if (bundleUrl != nil) {
      self.bundleUrl = bundleUrl;
    }
    
    if (zipBundleUrl != nil) {
     self.zipBundleUrl = zipBundleUrl;
    }
    
    if (bundlePath != nil) {
      self.bundlePath = bundlePath
    }
  }

  public func getBundleUrl()->String? {
    return bundleUrl
  }

  public func getZipBundleUrl()->String? {
    return zipBundleUrl
  }
  
  public func getBundlePath()->String? {
    return bundlePath
  }

  public func reload() {
    if (self.krakenBundlePlugin != nil) {
      self.channel?.invokeMethod("reload", arguments: nil)
    }
  }

  public func onAttach(krakenBundlePlugin:KrakenBundlePlugin, channel:FlutterMethodChannel) {
    self.krakenBundlePlugin = krakenBundlePlugin
    self.channel = channel
  }
}
